docs/proposals/20240327-enhancedlivenessprobe.md

```diff
@@ -0,0 +1,189 @@
+there are an enhanced livenessProbe webhook, an enhanced livenessProbe map nodePodProbe resource controller, 
+an enhanced livenessProbe controller and a node detection component.
+
+● **EnhancedLivenessProbeWebhook**: A webhook works in Kruise-manager. 
+The native livenessProbe is converted to the specific string configuration field in pod annotations as soon as the pod is created.
+
+● **EnhancedLivenessProbeMapNodePodProbeController**: It is a service component controller which is developed in Kruise-manager.
+The specific string configuration known as the livenessProbe is converted to the nodePodProbe(a custom resource in Kruise suite) resource immediately.
+
+● **EnhancedLivenessProbeController**: It is a service component controller which is developed in Kruise-manager.
+It watches the state change of the nodePodProbe resource and analyzes the application workload status with the failed livenessProbe pods. 
+This controller decide whether to restart containers or not(using the container recreate request, a custom resource in Kruise). 
+In calculating process, the resilience protection and application maxUnavailable protection policies are considered by the special algorithm and detection method comprehensively.
+
+● **LivenessProbeDetector**: It is just a single controller in Kruise-daemon, 
+which support to check the nodePodProbe(e.g., tcp/http/exec checking) as the same by the kubelet probe logic, using the node pod probe resource in Kruise.
+
+### Implementation
+
+**EnhancedLivenessProbeWebhook** 
+> When a pod created with livenessProbe config with `Annotations: apps.kruise.io/using-enhanced-liveness: true` feature gate, 
+the webhook will admit the pod resource to remove the native livenessProbe and patch the configuration to the pod annotations `apps.kruise.io/container-probe-context: '{...json format for liveness probe with specified contaienrs in pod...}''`.
+
+**EnhancedLivenessProbeMapNodePodProbeController**
+> For using the nodPodProbe controller processing, 
+> the livenessProbe config filed in the pod annotations should be converted to the nodePodProbe custom defined resource in Kruise suite. 
+> This controller can create and update the nodePodProbe resource.
+
+**EnhancedLivenessProbeController**
+> The overall architecture is shown in Figure 1 as below. 
+> After the enhanced livenessProbe controller watches the state change of node pod probe configured the failed livenessProbe pods, 
+> this decides immediately whether to perform the restart logic using the CRR(container recreate request) or not. 
+
+> In order to take into account the high availability of the application and service protection, 
+> the architecture design is coordinated with the PodUnavailableBudget function in Kruise suite and some anomaly detection algorithms(global switch, black list, global limiting rule, etc.). 
+> These protect the availability of application services before executing the container restart effectively, 
```

Comment:
plz explain in detail about the flow control logic

Reply:
A basic token bucket algorithm framework is released to control the restart frequency of the application with failed livenessProbe pods.